Sources/SPMBuildCore/ExtensionEvaluator.swift

```diff
@@ -0,0 +1,328 @@
+    public func evaluateExtensions(
+        buildEnvironment: BuildEnvironment,
+        execsDir: AbsolutePath,
+        outputDir: AbsolutePath,
+        extensionRunner: ExtensionRunner,
+        diagnostics: DiagnosticsEngine,
+        fileSystem: FileSystem
+    ) throws -> [ResolvedTarget: [ExtensionEvaluationResult]] {
+        // TODO: Convert this to be asynchronous, taking a completion closure.  This may require changes to package graph APIs.
+        var evalResultsByTarget: [ResolvedTarget: [ExtensionEvaluationResult]] = [:]
+        
+        for target in self.reachableTargets {
+            // Infer extensions from the declared dependencies, and collect them as well as any regular dependnencies.
+            // TODO: We'll want to separate out extension usages from dependencies, but for now we get them from dependencies.
+            var extensionTargets: [ExtensionTarget] = []
+            var dependencyTargets: [Target] = []
+            for dependency in target.dependencies(satisfying: buildEnvironment) {
+                switch dependency {
+                case .target(let target, _):
+                    if let extensionTarget = target.underlyingTarget as? ExtensionTarget {
+                        extensionTargets.append(extensionTarget)
+                    }
+                    else {
+                        dependencyTargets.append(target.underlyingTarget)
+                    }
+                case .product(_, _):
+                    // TODO: Support extension product dependencies.
+                    break
+                }
+            }
+            
+            // Leave quickly in the common case of not using any extensions.
+            if extensionTargets.isEmpty {
+                continue
+            }
+            
+            // If this target does use any extensions, create the input context to pass to them.
+            // FIXME: We'll want to decide on what directories to provide to the extenion
+            let package = self.packages.first{ $0.targets.contains(target) }!
```

Comment:
first! safe? replace with guard to make safer long term?

Reply:
Yes, this should be an internal error.  A leftover from before cleaning up the code.  Thanks for catching!

Reply:
Also, one would think that mapping a target back to its package should be common enough to have a method of some sort (which could then do caching etc) but I didn't spot one.  I can create one but wanted to keep the code minimal here.  But this should definitely throw an error (I try to avoid ! completely).